### PR TITLE
Native Z3 interface

### PR DESCRIPTION
**Pull request by manasij7479**
*.with.overflow encoding seems mysteriously broken.
Other things work, and is about 20-30% faster than our klee+smtlib interface. 
Don't merge yet.

### CODE REVIEW

**Comment by manasij7479** [comment]
If any of you want to look:
`./souper-check -souper-use-z3 filename.opt`
```
%x:i25 = var
%y:i25 = var
%z = sadd.with.overflow %x, %y
%ov = extractvalue %z, 1
pc %ov 0:i1
%r = extractvalue %z, 0
infer %r
%r2 = addnsw %x, %y
result %r2

```

**Comment by manasij7479** [comment]
Another weird bug, documenting here to pick up later
```
%A:i3 = var
%B:i3 = var
%X:i3 = var
%s = shl 1, %A
%Y = lshr %s, %B
%r = udiv %X, %Y
infer %r
%sub = sub %A, %B
%Y2 = shl 1, %sub
%r2 = udiv %X, %Y2
result %r2

```
z3 exprs generated:
```
LHS:(bvudiv %var_X (bvlshr (bvshl #b001 %var_A) %var_B))
RHS:(bvudiv %var_X (bvshl #b001 (bvsub %var_A %var_B)))
(define-fun %var_X () (_ BitVec 3)
  #b011)
(define-fun %var_A () (_ BitVec 3)
  #b101)
(define-fun %var_B () (_ BitVec 3)
  #b100)
  Invalid
```

**Comment by regehr** [comment]
nice!!
should I start to read and give feedback? I'm out hiking Sat but around other than that

**Comment by manasij7479** [comment]
No, wait until Monday maybe.
Also hiking tomorrow (Wheeler peak @Great Basin NP)

**Comment by manasij7479** [comment]
Another simple looking transform which doesn't verify.

```
%X:i8 = var
%Y:i8 = var
%c:i1 = var
%sel = select %c, %Y, 0
%r = udiv %X, %sel
infer %r
%r2 = udiv %X, %Y
result %r2
```

**Comment by regehr** [comment]
did you concatenate the operation result and the overflow bit in the correct order?

**Comment by regehr** [comment]
it might make you feel better to know that the problem with overflow intrinsics isn't related to path conditions, this can't be verified either:

```
%x:i25 = var
%y:i25 = var
%z = sadd.with.overflow %x, %y
%r = extractvalue %z, 0
infer %r
%r2 = add %x, %y
result %r2
```

**Comment by regehr** [comment]
ok so your add.with.overflow isn't computing the right result when there's overflow. 

remember: always simplify the circumstances surrounding a failing test.

```
%x:i4 = var
%y:i4 = var
%z = sadd.with.overflow 15:i4, 2:i4
%r = extractvalue %z, 0
infer %r
result 1:i4
``

**Comment by regehr** [comment]
this one does verify 

```
regehr@home:~/souper-manasij$ cat foo.opt
%X:i8 = var
%Y:i8 = var
%c:i1 = var
%sel = select %c, %Y, 0
%r = udiv %X, %sel
infer %r
%r2 = udiv %X, %Y
result %r2
regehr@home:~/souper-manasij$ ./build/souper-check foo.opt -souper-double-check 
; LGTM
regehr@home:~/souper-manasij$ ./build/souper-check foo.opt -souper-double-check -souper-use-z3
PC: #b1
LHS:(bvudiv %var_X (ite (= %var_c #b1) %var_Y #x00))
RHS:(bvudiv %var_X %var_Y)
; LGTM
regehr@home:~/souper-manasij$ 
```

**Comment by regehr** [comment]
so does this one ```regehr@home:~/souper-manasij$ cat foo.opt
%A:i3 = var
%B:i3 = var
%X:i3 = var
%s = shl 1, %A
%Y = lshr %s, %B
%r = udiv %X, %Y
infer %r
%sub = sub %A, %B
%Y2 = shl 1, %sub
%r2 = udiv %X, %Y2
result %r2
regehr@home:~/souper-manasij$ ./build/souper-check foo.opt -souper-double-check -souper-use-z3
PC: #b1
LHS:(bvudiv %var_X (bvlshr (bvshl #b001 %var_A) %var_B))
RHS:(bvudiv %var_X (bvshl #b001 (bvsub %var_A %var_B)))
; LGTM
regehr@home:~/souper-manasij$ ./build/souper-check foo.opt -souper-double-check 
; LGTM
regehr@home:~/souper-manasij$ 
```

**Comment by regehr** [comment]
`-souper-use-z3` isn't a very informative option since we already have 2 different ways to uses Z3 and you're adding a third, please change this a bit

**Comment by manasij7479** [comment]
concat order was wrong, everything seems to be working now.
This isn't integrated with synthesis yet, but all the solver checks pass.

**Comment by regehr** [comment]
ok, nice work!

I want a test case for timeouts using this API. do like a 1024 bit division or something like that where it has no chance whatsoever of completing.

have you tried turning off the KLEE code entirely? does everything still work in that case?

File: include/souper/Infer/Z3Driver.h [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    InstNumbers = 201;
+    //201 is chosen arbitrarily.
+
+    Translate(LHS);
+    if (Precondition) AddConstraint(Precondition);
+  }
+  bool verify(Inst *RHS, Inst *RHSAssumptions = nullptr) {
+    Translate(RHS);
+    if (RHSAssumptions) AddConstraint(RHSAssumptions);
```
Comment (regehr) [inline include/souper/Infer/Z3Driver.h:24]: put the "then" part on another line please

**Comment by regehr** [comment]
so @manasij7479 do you want to land this and then do models in a separate pr? if so, that's fine

I want you to support (and test) Z3 timeouts in this PR though

File: include/souper/Infer/Z3Driver.h [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    if (I->K == Inst::Kind::AShrExact) {
+      Solver.add(z3expr::ashr_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+    if (I->K == Inst::Kind::LShrExact) {
+      Solver.add(z3expr::lshr_exact(Get(I->Ops[0]), Get(I->Ops[1])));
+    }
+  }
+
+  bool Translate(souper::Inst *I) {
+    if (!I) return true;
```
Comment (regehr) [inline include/souper/Infer/Z3Driver.h:128]: on a new line

File: include/souper/Infer/Z3Driver.h [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+      if (!Translate(Op)) {
+        return false;
+      }
+      if (I->K == Inst::ExtractValue) {
+        break; // Break after translating main arg, idx is handled separately.
+      }
+
+    }
+
+    std::string Name = "";
```
Comment (regehr) [inline include/souper/Infer/Z3Driver.h:153]: remove unnecessary initializer

File: include/souper/Infer/Z3Driver.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      NamesCache[I] = Name;
+    }
+
+    auto W = I->Width;
+    addExtraPreds(I);
+    switch (I->K) {
+      case souper::Inst::Var: {
+        Put(I, ctx.bv_const(Name.c_str(), W));
+        auto DFC = getDataflowConditions(I, IC);
+        if (DFC) AddConstraint(DFC);
```
Comment (regehr) [inline include/souper/Infer/Z3Driver.h:176]: put the then part on a new line

File: include/souper/Infer/Z3Driver.h [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+      }
+      case souper::Inst::Const: {
+        Put(I, ctx.bv_val(I->Val.toString(10, false).c_str(), W));
+        // inefficient?
+        return true;
+      }
+
+      case souper::Inst::Phi: {
+        if (Ops.size() != 1) {
+          assert(false && "Phi with muliple arguments unimplemented");
```
Comment (regehr) [inline include/souper/Infer/Z3Driver.h:190]: plz fix this before we merge your patch

**Comment by regehr** [comment]
@manasij7479 can you explain what the changes to the alive driver are doing?

File: lib/Infer/Z3Driver.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#include "souper/Infer/Z3Driver.h"
+#include "souper/Infer/Z3Expr.h"
+
+extern unsigned DebugLevel;
+
+namespace souper {
+bool isTransformationValidZ3(souper::Inst *LHS, souper::Inst *RHS,
+                           const std::vector<InstMapping> &PCs,
```
Comment (regehr) [inline lib/Infer/Z3Driver.cpp:8]: fix indentation

File: include/souper/Infer/Z3Driver.h (hunk 5 above)
Comment (manasij7479) [reply include/souper/Infer/Z3Driver.h:190]: Phi's are handled at another level of abstraction. Both this and AliveDriver are responsible for verifying a transformation, given a precondition. Another layer decomposes LHS's with 'actual' phi nodes into multiple such (Pre => LHS == RHS?) problems.

**Comment by manasij7479** [comment]
AliveDriver changes are just splitting out reusable code into a separate TU.

**Comment by manasij7479** [comment]
> I want you to support (and test) Z3 timeouts in this PR though

Will do.

File: include/souper/Infer/Z3Driver.h (hunk 5 above)
Comment (regehr) [reply include/souper/Infer/Z3Driver.h:190]: ah, nice, thanks-- please add a comment saying phis have already been broken down before you get to this point

**Comment by manasij7479** [comment]
Implemented but not yet tested support for timeouts.
Even for larger bitwidths, z3 tactics seem to be taking care of solving things fast.
Have to find a weird valid transformation which escapes simplification.

**Comment by fvrmatteo** [comment]
@manasij7479 the "weird valid transformation which escapes simplification" could be here: https://github.com/google/souper/issues/738. Most likely the timeout will kick in before the synthesis candidate is verified as correct. I hope this helps and is not out of scope.

**Comment by regehr** [comment]
very wide division by non-constant should do it

**Comment by manasij7479** [comment]
See the newer PR

**Pull request closed without merging**